docs/wolfpasswd.1
'\" t
.\"     Title: wolframe
.\"    Author:  <contact@wolframe.net>
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 08/29/2014
.\"    Manual: Wolframe User Manual
.\"    Source: wolfpasswd
.\"  Language: English
.\"
.TH "WOLFRAME" "1" "08/29/2014" "wolfpasswd" "Wolframe User Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
wolfpasswd \- Wolframe password management program
.SH "SYNOPSIS"
.HP \w'\fBwolfpasswd\fR\ 'u
\fBwolfpasswd\fR \fB\-D\ \fR\fB\fIpasswordfile\fR\fR\fB\ \fR\fB\fIusername\fR\fR
.HP \w'\fBwolfpasswd\fR\ 'u
\fBwolfpasswd\fR [\fB\-c\fR] [\fB\-d \fR\fB\fIdays\fR\fR] \fIpasswordfile\fR \fIusername\fR [\fIuser\ info\fR]
.HP \w'\fBwolfpasswd\fR\ 'u
\fBwolfpasswd\fR \fB\-b\fR [\fB\-c\fR] [\fB\-d \fR\fB\fIdays\fR\fR] \fIpasswordfile\fR \fIusername\fR \fIpassword\fR [\fIuser\ info\fR]
.HP \w'\fBwolfpasswd\fR\ 'u
\fBwolfpasswd\fR \fB\-n\fR [\fB\-s \fR\fB\fIsalt\fR\fR] [\fB\-d \fR\fB\fIdays\fR\fR] \fIusername\fR [\fIuser\ info\fR]
.HP \w'\fBwolfpasswd\fR\ 'u
\fBwolfpasswd\fR \fB\-n\fR \fB\-b\fR [\fB\-s \fR\fB\fIsalt\fR\fR] [\fB\-d \fR\fB\fIdays\fR\fR] \fIusername\fR \fIpassword\fR [\fIuser\ info\fR]
.HP \w'\fBwolfpasswd\fR\ 'u
\fBwolfpasswd\fR \fB\-h\fR
.SH "DESCRIPTION"
.PP
This manual page documents briefly the
\fBwolfpasswd\fR
command\&.
.PP
\fBwolfpasswd\fR
is a tool to manage user credentials in the Wolframe system\&.
.SH "OPTIONS"
.PP
The program follows the usual GNU command line syntax, with long options starting with two dashes (`\-\*(Aq)\&. A summary of options is included below\&. For a complete description, see the
\fBinfo\fR(1)
files\&.
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Show summary of options\&.
.RE
.PP
\fB\-c\fR, \fB\-\-create\fR
.RS 4
Create the file if it doesn\*(Aqt exist\&.
.RE
.PP
\fB\-s\fR, \fB\-\-salt \fR\fB\fIsalt\fR\fR
.RS 4
Use the specified salt (only valid with \-\-display\-only)\&.
.RE
.PP
\fB\-d\fR, \fB\-\-days \fR\fB\fIdays\fR\fR
.RS 4
Set the password expiration after
\fIdays\fR
days\&. The default (=0) means don\*(Aqt expire\&.
.RE
.PP
\fB\-n\fR, \fB\-\-display\-only\fR
.RS 4
Don\*(Aqt update the password file; display results on stdout\&.
.RE
.PP
\fB\-b\fR, \fB\-\-batch\fR
.RS 4
Use the password from the command line instead of prompting for it\&.
.RE
.PP
\fB\-D\fR, \fB\-\-delete\fR
.RS 4
Delete the specified user\&.
.RE
.SH "DIAGNOSTICS"
.PP
The following diagnostics may be issued on
stderr:
.PP
password file \*(Aq\fIpasswordfile\fR\*(Aq does not exist\&.
.RS 4
The password file doesn\*(Aqt exist\&. Use \-c/\-\-create to create it first\&.
.RE
.PP
\fBwolfpasswd\fR
provides some return codes, that can be used in scripts:
.\" line length increase to cope w/ tbl weirdness
.ll +(\n(LLu * 62u / 100u)
.TS
ll.
\fICode\fR	\fIDiagnostic\fR
T{
\fB0\fR
T}	T{
Program exited successfully\&.
T}
T{
\fB1\fR
T}	T{
An error occurred\&.
T}
.TE
.\" line length decrease back to previous value
.ll -(\n(LLu * 62u / 100u)
.sp
.SH "BUGS"
.PP
Find issues and bugs on
Github
at
\m[blue]\fB\%https://github.com/Wolframe/Wolframe/issues\fR\m[]\&.
.SH "SEE ALSO"
.PP
\fBwolframed\fR(8)
.SH "AUTHOR"
.PP
 <\&contact@wolframe\&.net\&>
.RS 4
This manual was written by the Wolframe Team\&.
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2010-2014 Wolframe Team
.br
.PP
The authors grant permission to copy, distribute and/or modify this manual page under the terms of the GNU General Public License Version 3 or newer (GPLv3)\&.
.sp


debian/changelog
wolframe (0.0.1-1) unstable; urgency=low

  * Initial release

 -- Mihai Barbos <mihai.barbos@gmail.com>  Sun, 29 Aug 2010 16:00:00 +0200


src/modules/authentication/textfile/wolfpasswd.cpp
/************************************************************************

 Copyright (C) 2011 - 2014 Project Wolframe.
 All rights reserved.

 This file is part of Project Wolframe.

 Commercial Usage
    Licensees holding valid Project Wolframe Commercial licenses may
    use this file in accordance with the Project Wolframe
    Commercial License Agreement provided with the Software or,
    alternatively, in accordance with the terms contained
    in a written agreement between the licensee and Project Wolframe.

 GNU General Public License Usage
    Alternatively, you can redistribute this file and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Wolframe is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Wolframe.  If not, see <http://www.gnu.org/licenses/>.

 If you have questions regarding the use of this file, please contact
 Project Wolframe.

************************************************************************/
//
// wolfpasswd - text file password utility
//

#include <boost/program_options.hpp>
#include <string>
#include <iostream>

#include "utils/getPassword.hpp"
#include "utils/fileUtils.hpp"
#include "passwdFile.hpp"
#include "AAAA/passwordHash.hpp"
#include "types/base64.hpp"
#include "system/globalRngGen.hpp"
#define BOOST_FILESYSTEM_VERSION 3
#include <boost/filesystem.hpp>

namespace PO = boost::program_options;
namespace WA = _Wolframe::AAAA;
namespace WU = _Wolframe::utils;


int main( int argc, char* argv[] )
{
	unsigned	days;
	std::string	saltStr;
	std::string	b64SaltStr;

	PO::options_description desc( "Usage:\n"
				      "  wolfpasswd -D passwordfile username\n"
				      "  wolfpasswd [-c] [-d days] passwordfile username [user info]\n"
				      "  wolfpasswd -b[c] [-d days] passwordfile username password [user info]\n"
				      "  wolfpasswd -n [-s/-S salt] [-d days] username [user info]\n"
				      "  wolfpasswd -nb [-s/-S salt] [-d days] username password [user info]\n"
				      "  wolfpasswd -h\n"
				      "Options" );
	desc.add_options()
			( "help,h", "Display this help message." )
			( "create,c", "Create the file if it doesn't exist." )
			( "salt,s", PO::value< std::string >( &saltStr ), "Use the specified plain text salt (only valid with --display-only)." )
			( "base64-salt,S", PO::value< std::string >( &b64SaltStr ), "Use the specified base64 salt (only valid with --display-only)." )
			( "days,d", PO::value< unsigned >( &days )->default_value( 0 ), "Set the password expiration after <days>." )
			( "display-only,n", "Don't update the password file; display results on stdout." )
			( "batch,b", "Use the password from the command line instead of prompting for it.")
			( "delete,D", "Delete the specified user." )
			;

	PO::options_description cmdArgs( "Arguments" );
	cmdArgs.add_options()
			( "posArgs", PO::value< std::vector<std::string> >(), "positional arguments" )
			;
	cmdArgs.add( desc );

	PO::positional_options_description posArgs;
	posArgs.add( "posArgs", -1 );

	PO::variables_map vm;
	try	{
		PO::store( PO::command_line_parser( argc, argv ).
			   options( cmdArgs ).positional( posArgs ).run(), vm );
		PO::notify( vm );
	}
	catch( std::exception& e )	{
		std::cerr << "\nERROR: " << e.what() << "\n\n";
		std::cout << desc << "\n";
		return 2;
	}

	bool	createFile = false;
	bool	displayOnly = false;
	bool	batchPwd = false;
	bool	delUser = false;

	if ( vm.count( "create" ))
		createFile = true;
	if ( vm.count( "display-only" ))
		displayOnly = true;
	if ( vm.count( "batch" ))
		batchPwd = true;
	if ( vm.count( "delete" ))
		delUser = true;

	// deal with help
	if ( vm.count( "help" ))	{
		if ( createFile || displayOnly|| batchPwd || delUser
				|| vm.count( "posArgs" ))
			std::cout << "\nWARNING: --help ignores all other flags and arguments.\n\n";
		std::cout << desc << "\n";
		return 1;
	}

	if ( !vm.count( "posArgs" ))	{
		std::cerr << "\nERROR: no arguments given.\n\n";
		std::cout << desc << "\n";
		return 2;
	}
	const std::vector<std::string>& args = vm["posArgs"].as< std::vector<std::string> >();

	_Wolframe::GlobalRandomGenerator& rnd = _Wolframe::GlobalRandomGenerator::instance( "" );

	// display only
	if ( displayOnly )	{
		bool	wrong = false;
		if ( createFile )	{
			std::cerr << "\nERROR: --create cannot be used with --display-only.";
			wrong = true;
		}
		if ( delUser )	{
			std::cerr << "\nERROR: --delete cannot be used with --display-only.";
			wrong = true;
		}
		if ( batchPwd )	{
			if ( args.size() < 2 )	{
				std::cerr << "\nERROR: too few arguments.";
				wrong = true;
			}
			else if ( args.size() > 3 )	{
				std::cerr << "\nERROR: too many arguments.";
				wrong = true;
			}
		}
		else	{
			if ( args.size() > 2 )	{
				std::cerr << "\nERROR: too many arguments.";
				wrong = true;
			}
		}
		if ( !saltStr.empty() && !b64SaltStr.empty() )	{
			std::cerr << "\nERROR: both --salt and --base64-salt arguments given.";
			wrong = true;
		}
		if ( wrong )	{
			std::cout << "\n\n" << desc << "\n";
			return 2;
		}

		// All parameters are OK
		std::string passwd;
		WA::PwdFileUser user;
		if ( ! batchPwd )	{
			passwd = WA::getPassword();
			if( args.size() == 2 )
				user.info = args[1];
		}
		else	{
			passwd = args[1];
			if( args.size() == 3 )
				user.info = args[2];
		}
		// now do the job
		WA::PasswordHash::Salt salt;
		if ( saltStr.empty() && b64SaltStr.empty() )
			salt = WA::PasswordHash::Salt( rnd );
		else	{
			if ( !saltStr.empty() )
				b64SaltStr = _Wolframe::base64::encode( saltStr.data(), saltStr.size(), 0 );
			salt = WA::PasswordHash::Salt( b64SaltStr );
		}
		WA::PasswordHash pwd( salt, passwd );

		user.user = args[0];
		user.hash = pwd.toString();
		user.expiry = 0;
		std::cout << "Password line: " << WA::PasswordFile::passwdLine( user );
	}
	// delete user
	else if ( delUser )	{
		bool	wrong = false;
		if ( createFile )	{
			std::cerr << "\nERROR: --create cannot be used with --delete.";
			wrong = true;
		}
		if ( args.size() < 2 )	{
			std::cerr << "\nERROR: too few arguments.";
			wrong = true;
		}
		else if ( args.size() > 2 )	{
			std::cerr << "\nERROR: too many arguments.";
			wrong = true;
		}
		if ( wrong )	{
			std::cout << "\n\n" << desc << "\n";
			return 2;
		}
		if ( ! saltStr.empty() )
			std::cerr << "\nWarning: --salt argument ignored.";
		if ( ! b64SaltStr.empty() )
			std::cerr << "\nWarning: --base64-salt argument ignored.";

		// All parameters are OK
		try	{
			std::string filename = WU::resolvePath( boost::filesystem::absolute( args[0] ).string());
			WA::PasswordFile pwdFile( filename );

			if ( pwdFile.delUser( args[1] ))
				std::cout << "User '" << args[1] << "' removed from password file '"
					  << filename << "'.";
			else
				std::cout << "User '" << args[1] << "' not found in password file '"
					  << filename << "'.";
		}
		catch( std::exception& e )	{
			std::cerr << "Error removing user '" << args[1] << "': " << e.what() << "\n\n";
			return 1;
		}
	}
	// normal operation
	else	{
		bool	wrong = false;
		if ( batchPwd )	{
			if ( args.size() < 3 )	{
				std::cerr << "\nERROR: too few arguments.";
				wrong = true;
			}
			else if ( args.size() > 4 )	{
				std::cerr << "\nERROR: too many arguments.";
				wrong = true;
			}
		}
		else	{
			if ( args.size() < 2 )	{
				std::cerr << "\nERROR: too few arguments.";
				wrong = true;
			}
			if ( args.size() > 3 )	{
				std::cerr << "\nERROR: too many arguments.";
				wrong = true;
			}
		}
		if ( !saltStr.empty() && !b64SaltStr.empty() )	{
			std::cerr << "\nERROR: both --salt and --base64-salt arguments given.";
			wrong = true;
		}

		if ( wrong )	{
			std::cout << "\n\n" << desc << "\n";
			return 2;
		}

		// Al parameters are OK, do the job
		bool create = false;
		if ( createFile )
			create = true;
		try	{
			std::string filename = WU::resolvePath( boost::filesystem::absolute( args[0] ).string());
			WA::PasswordFile pwdFile( filename, create );
			WA::PwdFileUser user;

			std::string passwd;
			if ( ! batchPwd )	{
				passwd = WA::getPassword();
				if( args.size() == 3 )
					user.info = args[2];
			}
			else	{
				passwd = args[2];
				if( args.size() == 4 )
					user.info = args[3];
			}
			WA::PasswordHash::Salt salt;
			if ( saltStr.empty() && b64SaltStr.empty() )
				salt = WA::PasswordHash::Salt( rnd );
			else	{
				if ( !saltStr.empty() )
					b64SaltStr = _Wolframe::base64::encode( saltStr.data(), saltStr.size(), 0 );
				salt = WA::PasswordHash::Salt( b64SaltStr );
			}
			WA::PasswordHash pwd( salt, passwd );
			user.user = args[1];
			user.hash = pwd.toString();
			user.expiry = 0;
			if ( pwdFile.addUser( user ) )
				std::cout << "User '" << args[1] << "' added to the password file '"
					  << filename << "'.";
			else
				std::cout << "Error: user '" << args[1] << "' already exists in the password file '"
					  << filename << "'.";

		}
		catch( std::exception& e )	{
			std::cerr << "Error adding user '" << args[1] << "': " << e.what() << "\n\n";
			return 1;
		}
	}
	std::cout << "\nDone.\n\n";
	return 0;
}
